dag-json/src/lib.rs

```diff
@@ -0,0 +1,236 @@
+        fmt.write_str("any valid JSON value")
+    }
+
+    #[inline]
+    fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        self.visit_string(String::from(value))
+    }
+
+    #[inline]
+    fn visit_string<E>(self, value: String) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        Ok(Ipld::String(value))
+    }
+    #[inline]
+    fn visit_bytes<E>(self, v: &[u8]) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        self.visit_byte_buf(v.to_owned())
+    }
+
+    #[inline]
+    fn visit_byte_buf<E>(self, v: Vec<u8>) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        Ok(Ipld::Bytes(v))
+    }
+
+    #[inline]
+    fn visit_u64<E>(self, v: u64) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        Ok(Ipld::Integer(v.into()))
+    }
+
+    #[inline]
+    fn visit_i64<E>(self, v: i64) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        Ok(Ipld::Integer(v.into()))
+    }
+
+    #[inline]
+    fn visit_i128<E>(self, v: i128) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        Ok(Ipld::Integer(v))
+    }
+
+    #[inline]
+    fn visit_bool<E>(self, v: bool) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        Ok(Ipld::Bool(v))
+    }
+
+    #[inline]
+    fn visit_none<E>(self) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        self.visit_unit()
+    }
+
+    #[inline]
+    fn visit_unit<E>(self) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        Ok(Ipld::Null)
+    }
+
+    #[inline]
+    fn visit_seq<V>(self, mut visitor: V) -> Result<Self::Value, V::Error>
+    where
+        V: de::SeqAccess<'de>,
+    {
+        let mut vec: Vec<WrapperOwned> = Vec::new();
+
+        while let Some(elem) = visitor.next_element()? {
+            vec.push(elem);
+        }
+
+        let unwrapped = vec.into_iter().map(|WrapperOwned(ipld)| ipld).collect();
+        Ok(Ipld::List(unwrapped))
+    }
+
+    #[inline]
+    fn visit_map<V>(self, mut visitor: V) -> Result<Self::Value, V::Error>
+    where
+        V: de::MapAccess<'de>,
+    {
+        let mut values: Vec<(String, WrapperOwned)> = Vec::new();
+
+        while let Some((key, value)) = visitor.next_entry()? {
+            values.push((key, value));
+        }
+
+        // JSON Object represents IPLD Link if it is `{ "/": "...." }` therefor
+        // we valiadet if that is the case here.
+        if let Some((key, WrapperOwned(Ipld::String(value)))) = values.first() {
+            if key == LINK_KEY && values.len() == 1 {
+                let link = base64::decode(value);
+                // TODO: Find out what is the expected behavior in cases where
+                // value is not a valid CID (or base64 endode string here). For
+                // now treat it as some other JSON Object.
```

Comment:
I think it should just fail. Treating `{ "/": "Qm1234" }` as a link and `{ "/": "hello" }` as a normal JSON object smells like DWIM.